subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectContainerExtensions.kt

```diff
@@ -66,28 +66,28 @@ class NamedDomainObjectContainerScope<T : Any>(
         polymorphicDomainObjectContainer().containerWithType(type)
 
     /**
-     * @see [NamedDomainObjectContainer.maybeCreate]
+     * @see [NamedDomainObjectContainer.named]
      */
-    inline operator fun String.invoke(configuration: T.() -> Unit): T =
-        this().apply(configuration)
+    operator fun String.invoke(configuration: T.() -> Unit): DomainObjectProvider<T> =
+        this().also { it.configure(configuration) }
 
     /**
-     * @see [NamedDomainObjectContainer.maybeCreate]
+     * @see [NamedDomainObjectContainer.named]
      */
-    operator fun String.invoke(): T =
-        container.maybeCreate(this)
+    operator fun String.invoke(): DomainObjectProvider<T> =
+        container.named(this)
 
     /**
-     * @see [PolymorphicDomainObjectContainer.maybeCreate]
+     * @see [PolymorphicDomainObjectContainer.named]
      */
-    inline operator fun <U : T> String.invoke(type: KClass<U>, configuration: U.() -> Unit): U =
-        this(type).apply(configuration)
+    operator fun <U : T> String.invoke(type: KClass<U>, configuration: U.() -> Unit): DomainObjectProvider<U> =
+        this(type).also { it.configure(configuration) }
 
     /**
-     * @see [PolymorphicDomainObjectContainer.maybeCreate]
+     * @see [PolymorphicDomainObjectContainer.named]
      */
-    operator fun <U : T> String.invoke(type: KClass<U>): U =
-        polymorphicDomainObjectContainer().maybeCreate(this, type.java)
+    operator fun <U : T> String.invoke(type: KClass<U>): DomainObjectProvider<U> =
+        polymorphicDomainObjectContainer().withType(type.java).named(this)
```

Comment:
:x: The semantics should be that the caller is asserting the type of the named element, not querying elements of the specific type, similar to [`NamedDomainobjectCollection.getByName<T>(String)`](https://github.com/gradle/kotlin-dsl/blob/448edf838543abb886968bab59e9bc28f1c6111b/subprojects/provider/src/main/kotlin/org/gradle/kotlin/dsl/NamedDomainObjectCollectionExtensions.kt#L60).

Reply:
The assertion would need to be lazy to not realize the container element, as a configuration block maybe. Then the assertion happens sometime in the future, when/if the element is realized.

Reply:
I agree the semantics are different. If the targeted element cannot be found (by type and name) you'll get a "not found" exception. What you'd not get is the detail about an element found with the requested name but of a different type.

Reply:
Yes, the assertion should happen during configuration.

Reply:
Then it requires an eager cast to return the requested provider type that may fail before we have a chance to provide a helpful error:
```kotlin
operator fun <U : T> String.invoke(type: KClass<U>): DomainObjectProvider<U> =
    polymorphicDomainObjectContainer().named(this).apply {
        configure {
            it as? U
                ?: illegalElementType(container, this@invoke, type, it::class)
        }
    } as DomainObjectProvider<U>
```

Or did I miss something?

Reply:
I was thinking on something like:

```kotlin
    /**
     * @see [PolymorphicDomainObjectContainer.named]
     */
    operator fun <U : T> String.invoke(type: KClass<U>, configuration: U.() -> Unit): DomainObjectProvider<U> =
        uncheckedCast(
            polymorphicDomainObjectContainer().named(this).apply {
                configure {
                    configuration(
                        type.safeCast(it)
                            ?: throw illegalElementType(container, name, type, it::class)
                    )
                }
            })
```

Reply:
Yeah, it's basically the same.
Wouldn't the unchecked cast fail earlier than the type check?